torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+import copy
+from enum import Enum, auto
+import functools
+import traceback
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+import torch
+from torch.autograd import Variable
+import torch.distributed as dist
+from torch.distributed import ProcessGroup
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.parameter import Parameter
+
+from torch.distributed.distributed_c10d import _get_default_group
+
+from .utils import (
+    _apply_to_tensors,
+)
+
+if TYPE_CHECKING:
+    from collections import OrderedDict  # noqa: F401
+
+
+class TrainingState(Enum):
+    """
+    Simple enum to indicate what state FSDP is in. Used for asserting
+    to make sure APIs are called in the correct state.
+    ..note::
+        BACKWARD_PRE and BACKWARD_POST states are used to ensure we
+        receives backward hooks in the correct order. It is used to catch
+        unexpected order of hooks being called (likely due to our
+        hook registration logic or autograd engine logic changes).
+    """
+
+    IDLE = auto()
+    FORWARD = auto()
+    BACKWARD_PRE = auto()
+    BACKWARD_POST = auto()
+
+
+class FullyShardedDataParallel(nn.Module):
+    """
+    A wrapper for sharding Module parameters across data parallel workers. This
+    is inspired by `Xu et al.`_ as well as the ZeRO Stage 3 from DeepSpeed_.
+    FullyShardedDataParallel is commonly shorten to FSDP.
+    .. _`Xu et al.`: https://arxiv.org/abs/2004.13336
+    .. _DeepSpeed: https://www.deepspeed.ai/
+    Example::
+        import torch
+        from torch.distributed._fsdp import FullyShardedDataParallel as FSDP
```

Comment:
Do you want to actually call it FSDP if that is the recommended name for the users?

Reply:
It is a stylistic choice.
I just wanted to point out that if even the example has to use a shorter name, maybe the name should be short to begin with.

But not a blocker for sure!